lib/model/recent_senders.dart

```diff
@@ -0,0 +1,167 @@
+  void handleMessages(List<Message> messages, {bool inReverse = false}) {
+    for (int i = inReverse ? messages.length - 1 : 0;
+                 inReverse ? i >= 0              : i < messages.length;
+                 inReverse ? i--                 : i++) {
+      handleMessage(messages[i]);
+    }
+  }
+
+  /// Records the necessary data from [message] if it is a [StreamMessage].
+  ///
+  /// If [message] is not a [StreamMessage], this is a no-op.
+  void handleMessage(Message message) {
+    if (message is! StreamMessage) return;
+
+    final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+
+    // Track message in stream.
+    var sendersMap = streamSenders[streamId] ??= {};
+    var idTracker = sendersMap[senderId] ??= MessageIdTracker();
+    idTracker.add(messageId);
+
+    // Track message in topic.
+    final topicsMap = topicSenders[streamId] ??= {};
+    sendersMap = topicsMap[topic] ??= {};
+    idTracker = sendersMap[senderId] ??= MessageIdTracker();
+    idTracker.add(messageId);
+  }
+
+  void handleDeleteMessageEvent(DeleteMessageEvent event, Map<int, Message> cachedMessages) {
+    if (event.messageType != MessageType.stream) return;
+
+    for (final id in event.messageIds) {
+      final message = cachedMessages[id] as StreamMessage?;
+      if (message == null) break;
+      final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+
+    _removeMessageInStream(
+      streamId: streamId, senderId: senderId, messageId: messageId);
+
+    _removeMessageInTopic(streamId: streamId, topic: topic, senderId: senderId,
+      messageId: messageId);
+    }
+  }
+
+  void _removeMessageInStream({
+    required int streamId,
+    required int senderId,
+    required int messageId,
+  }) {
+    if (streamSenders.isEmpty) return;
+
+    final sendersMap = streamSenders[streamId];
+    if (sendersMap == null) return;
+
+    final idTracker = sendersMap[senderId];
+    if (idTracker == null) return;
+
+    idTracker.remove(messageId);
+    if (idTracker.maxId == null) sendersMap.remove(senderId);
+    if (sendersMap.isEmpty) streamSenders.remove(streamId);
+  }
+
+  void _removeMessageInTopic({
+    required int streamId,
+    required String topic,
+    required int senderId,
+    required int messageId,
+  }) {
+    if (topicSenders.isEmpty) return;
+
+    final topicsMap = topicSenders[streamId];
+    if (topicsMap == null) return;
+
+    final sendersMap = topicsMap[topic];
+    if (sendersMap == null) return;
+
+    final idTracker = sendersMap[senderId];
+    if (idTracker == null) return;
+
+    idTracker.remove(messageId);
+    if (idTracker.maxId == null) sendersMap.remove(senderId);
+    if (sendersMap.isEmpty) topicsMap.remove(topic);
+    if (topicsMap.isEmpty) topicSenders.remove(streamId);
+  }
+}
+
+class MessageIdTracker {
+  /// A list of distinct message IDs, sorted ascendingly.
+  final QueueList<int> _ids = QueueList.from([]);
```

Comment:
I think using `QueueList` is the best data structure for this scenario. It has $O(1)$ time complexity for both `addFirst` and `addLast`, as well as when accessing elements by index.